ujt/server/server.py

```diff
@@ -57,14 +64,46 @@ def GetSLIs(self, request, context):
         In this mock example, we generate new random values for SLIs dynamically.
         In a real server, this method should report real SLI values.
         """
+        current_timestamp = Timestamp()
+        current_timestamp.GetCurrentTime()
+        # Set the start and end time to the current time if they are unset in the request
+        start_dt = (
+            request.start_time.ToDatetime()
+            if request.start_time.IsInitialized()
+            else current_timestamp.ToDatetime()
+        )
+        end_dt = (
+            request.end_time.ToDatetime()
+            if request.end_time.IsInitialized()
+            else current_timestamp.ToDatetime()
+        )
 
-        slis = generate_new_random_slis(self.nodes)
-        # Update the server's internal Node state
-        sli_name_map = {sli.node_name: sli for sli in slis}
-        for node in self.nodes:
-            del node.slis[:]
-            node.slis.extend([sli_name_map[node.name]])
-        return server_pb2.GetSLIsResponse(slis=slis)
+        requested_node_names = (
+            request.node_names
+            if request.node_names != []
+            else [node.name for node in self.nodes]
+        )
+        slis_output = []
+        slis_at_timestamp = []
+        # In this implementation, the timestamps are inclusive since we generate new SLIs
+        # However, this is not inherent to the proto semantics
+        while start_dt <= end_dt:
+            slis_at_timestamp = generate_new_random_slis(requested_node_names, start_dt)
+            slis_output += slis_at_timestamp
+            start_dt += datetime.timedelta(seconds=self.reporting_interval)
+
+        # Can we improve this logic? Doesn't seem very elegant.
+        if slis_at_timestamp != []:
```

Comment:
You can clean some of this up at least:
slis_at_timestamp is only empty if end_time < start_time, which is a bad request anyway. Just validate that we're not in that case up front. this_last_reported_timestamp isn't really needed, you can just inline. You already have the current timestamp stored; it's start_dt since that's what you iterate, so you don't have to access the map.

Reply:
Agreed on making the interface contract more explicit via comments and explicit checking.

At the end of the loop, start_dt will be the last_reported_timestamp + the timedelta of reporting interval (rather than the last stored timestamp). Do you think it's better to subtract the timedelta?